### ISSUE TITLE
constant hashexpr not found in subplan's targetlist

### ISSUE THREAD

**Issue by liruto** [greenplum-db/gpdb#378]
A case with an internal error:
```
create table t1(tc1 int, tc2 int) distributed by (tc1);
mycluster=# explain select * from t1 t1  where tc1 in (select 1 from t1 t2 where tc1 in (select 1 from t1 t3 where t3.tc2 = t2.tc2));
ERROR:  Unexpected internal error (cdbpathtoplan.c:173)
```

In function '_cdbpath_match_preds_to_partkey_tail_', PathKey _copathkey_'s equivalence class could represent a pseudo-constant. If this happens, _colocus.partkey_h_ would contain a PathKey whose equivalence class is constant. But, '_cdbpullup_findPathKeyExprInTargetList_' doesn't take it into consideration. While PathKey _item_ equals constant at the same time as _targetlist_ is NIL, a NULL is returned, and then an error occurs.

After the following modification, it seems to run well.
```
--- a/src/backend/cdb/cdbpullup.c
+++ b/src/backend/cdb/cdbpullup.c
@@ -380,7 +380,7 @@ cdbpullup_findPathKeyExprInTargetList(PathKey *item, List *targetlist)
        EquivalenceClass *eclass = item->pk_eclass;
 
        /* Bail if targetlist is empty. */
-       if (!targetlist)
+       if (!targetlist && !CdbEquivClassIsConstant(eclass))
                return NULL;
 
        foreach(lc, eclass->ec_members)
@@ -392,6 +392,9 @@ cdbpullup_findPathKeyExprInTargetList(PathKey *item, List *targetlist)
                if (em->em_is_const)
                        return key;
 
+              if (CdbEquivClassIsConstant(eclass))
+                      continue;
+
                /* Ignore possible RelabelType node atop the PathKey expr. */
                if (IsA(key, RelabelType))
                        key = ((RelabelType *)key)->arg;
```

I'm not sure it, and it seems to be coming from the equivalence class merge patch.

**Comment by liruto** [comment]
@hlinnaka Could you take a look at this?

**Comment by hlinnaka** [comment]
Thanks, opened PR 404 to fix this. I moved code around the function a bit more, to make the code easier to understand, but it's basically the same as your modification above

**Comment by hlinnaka** [comment]
Fixed, thanks for the report! (I used wrong issue number in the commit, oopsie. The commit that fixed this was 81b5a62bd2ee21586494cbb10f051248cc363c14)

**Comment by danielgustafsson** [comment]
Closing as it has been fixed.
